### ISSUE TITLE
Please add a way to limit GUI refresh rate

### ISSUE THREAD

**Issue by Efenstor** [lsp-plugins/lsp-plugins#369]
Thank you for the amazing plugin set! I use Debian stable (bookworm) with the latest Ardour. My problem is that GUI overloads the CPU, so I can barely use any plugin at all. If I disable vsync (TearFree) in the xorg settings and turn off the analysis then the responsiveness becomes better but still far from perfect. The responsiveness of the plugin GUI itself is good but Ardour at the background becomes very slow. I use the hi-dpi monitor, so the interface is scaled to 200%, and I guess it makes things even worse. I have this problem on both of my machines with any LSP plugin. Sometimes it takes 20-30 seconds for Ardour just to respond to me pressing the "stop" button (or the space key) if the analyzer is on, if it's off then it's useable but obviously could be better (for example none of the Calf plugins has such problems). I think the solution is to limit the GUI refresh rate, may be as a manual option.

**Comment by sadko4u** [comment]
Hello! The refresh rate of GUI is limited to 25 FPS.
LSP uses cairo for rendering which is pretty OK dealing with hardware resources.
I need more information about your system configuration.
You also may have issues with your graphics driver.

**Comment by Efenstor** [comment]
I did some more tests and the video card may be a problem indeed, I'll definitely do more testing. My video card is MSI Radeon RX 5600 XT OC with the open-source amdgpu driver (v23.0.0, 20230210-5 firmware). There were some problems with it in the past (random black screens/system reboots) but with the 6.5.10 kernel it finally seems stable. It also caused sound stuttering with my old sound card (PCI MIA Echo), which may or may not be related, but those were gone since I swapped the sound card. In general LSP plugins are now the only piece of software which definitely has some weird issues, Kdenlive (which also uses Cairo), for example, works fine.

**Comment by Efenstor** [comment]
Now I think the video card is not a problem, it's more to do with process/thread prioritization. The thing is that not the whole system becomes sluggish but only the Ardour GUI: the LSP plugin GUI is updated fine, I can run any other program and it works perfectly. My CPU is AMD Ryzen 7 2700, it's 8-core but has 16 virtual cores. If I set in the Ardour preferences the number of signal processing threads to 7 or 8 and the CPU DMA latency setting to 400 usec it all becomes much less laggy. Too bad only the LSP plugins suffer from this, is there a way to lower their process priority?

**Comment by sadko4u** [comment]
The host sets the process and thread priority to plugins. LSP Plugins do nothing with this. All created threads are of default priority. If you use LV2 versions, the UI is handled by the host UI thread, see [lv2:IdleInterface](https://lv2plug.in/ns/extensions/ui#idleInterface).
I think it is a good reason to set up the `perf` tool and watch what's happening on your system by launching `perf top`.

**Comment by Efenstor** [comment]
Not sure what to make of it.

Here is "perf top --sort comm,dso" when stopped:
![Screenshot at 2023-12-30 22-41-39](https://github.com/lsp-plugins/lsp-plugins/assets/11175574/da4cfa37-86da-49a0-88d0-f699369b73c1)

When playing:
![Screenshot at 2023-12-30 22-41-13](https://github.com/lsp-plugins/lsp-plugins/assets/11175574/563daf13-da93-47ab-826e-adecaf6967f3)

and just "perf top" when playing:
![Screenshot at 2023-12-30 22-40-56](https://github.com/lsp-plugins/lsp-plugins/assets/11175574/e9775533-a252-4b23-bab2-b922a59753c7)

It is visible that the Ardour GUI is not redrawn for many seconds after the LSP window has been moved but the whole audio engine keeps playing perfectly.

**Comment by Efenstor** [comment]
Here is how it looks without any LSP plugin in the project:
![Uploading Screenshot at 2023-12-30 22-50-20.png…]()

**Comment by Efenstor** [comment]
Here is how it looks without any LSP plugin in the project:

![Screenshot at 2023-12-30 22-50-20](https://github.com/lsp-plugins/lsp-plugins/assets/11175574/4eecae98-9274-4ce8-9816-52cb653dcf6d)

**Comment by Efenstor** [comment]
I think now I see, it's really something with libcairo and my system because Calf does not use Cairo to draw anything, as well as any other plugin I use. Somehow it overloads the system. Not sure how to fix it.

**Comment by sadko4u** [comment]
I see the `swapper` process is eating too much CPU. It seems that you're having troubles with RAM and memory pages are constantly falling into drive's swap partition.

**Comment by Efenstor** [comment]
I don't think that's the case:
![Screenshot at 2023-12-31 00-07-47](https://github.com/lsp-plugins/lsp-plugins/assets/11175574/5374b58e-c8e7-405a-9cf4-27d7a636b008)

I can't even imagine why swapper may be so active. The only special thing about my swap partition is that it's LUKS-encrypted but as there's plenty of free RAM that should not be the case. Anyway I'll try to find out what's causing swapper to be so CPU-intensive.

To test out the bad video card theory I've dug out my old Radeon Aero, installed it and nope: the problem is still present, so it's unlikely the X11-driver-libcairo problem.

**Comment by sadko4u** [comment]
I have Ryzen 4800H CPU with embedded graphics and all works very well. Many other people reported that UI works pretty well. You're the only one who has such problem. I have no idea how to help you at this moment.

**Comment by Efenstor** [comment]
Solved it:

1. echo "vm.swappiness = 10" >> /etc/sysctl.conf
2. Ardour > Preferences > Performance > Power Management, CPU DMA Latency = Unset
3. (optional, recommended) BIOS > AMD CBS (CPU settings) > C-State (Global C-State Control) = Disabled

Thank you again!

**Comment by sadko4u** [comment]
@Efenstor 
Is there some recommendation you've found in internet?

**Comment by Efenstor** [comment]
Not quite: first I tried to deal with the swapper problem, here is the discussion on how to lower swappiness when there is enough of memory: https://unix.stackexchange.com/questions/265713/how-to-configure-swappiness-in-linux-memory-management. The second problem was the Ardour's clumsy integrated system of preventing modern CPUs (e.g. Ryzen) from going into the lowered power states (C-states), which also overloaded the swapper by continuously "pumping" the CPU; therefore I've deducted that it's much better to disable it in the BIOS completely.

**Comment by Efenstor** [comment]
I sorta reopen this issue, because a related problem still remains: when working on a complex project with lots and lots of processing and more than 10 tracks and if an LSP plugin with analysis display is shown the refresh rate of the Ardour interface drops to about 1 fps. I mean, is the analyzer so important to be displayed with perfect frame rate, while the whole interface in the background becomes sluggish? Of course I can turn off the analyzer completely but it's a useful thing after all.

My suggestion is to add:

1. A switch to draw the graphic display elements (analyzer and/or inline display) without anti-aliasing and using thin 1px lines ("low CPU mode").
2. If possible, an automatic adaptive refresh rate limit for the analyzer, deducing the value from the current GUI/DSP load or just the overall CPU load (I don't know whether it is possible or not)
3. If not possible, then at least a way to set the limit manually

**Comment by derkrasseleo** [comment]
I know this is the opposite of the problem described, but in general some option to adjust the frame rate would indeed be nice. Sometimes I feel like a higher refresh rate (in the analyzer plugins, for example) just makes sense, especially on high refresh rate displays.

**Comment by sadko4u** [comment]
The main problem with `libcairo` is that it doesn't like to draw many line segments. I noticed that when worked on the Oscilloscope plugin.
@derkrasseleo there is also difference between the UI refresh rate and the data synchronization rate between DSP and UI. Just increasing the UI refresh rate probably won't give the increase of data synchronization rate and will require more engineering.

**Comment by wargreen** [comment]
Hi,
I have a similar issue, with very slow refresh rate when 2 plugins GUI (or more) are open.
After some talks with the ardour's team, it seem linked with opengl in many child windows...
I don't know if this can be usefull, so my 0.2 cents :)

**Comment by brummer10** [comment]
I dont know If you use already, but cairo_push and cairo_pop makes a huge Differenz.

**Comment by sadko4u** [comment]
@brummer10 how should it help? Do you mean that `cairo_push_group` and `cairo_pop_group` should accelerate line drawing calls `cairo_move_to` and `cairo_line_to`?

**Comment by sadko4u** [comment]
I've just updated cairo surfaces to use `cairo_surface_create_similar` instead of `cairo_image_surface_create`. Also I changed antialiasing mode from GOOD to FAST and pushed changes to `devel` branch.
@Efenstor @wargreen @derkrasseleo are you available to build from `devel` branch and test if something has changed?

**Comment by brummer10** [comment]
> @brummer10 how should it help? Do you mean that `cairo_push_group` and `cairo_pop_group` should accelerate line drawing calls `cairo_move_to` and `cairo_line_to`?

It minimize the calls to th Graphik Server.

**Comment by Efenstor** [comment]
> @Efenstor @wargreen @derkrasseleo are you available to build from `devel` branch and test if something has changed?

I've built it now (v1.2.16); may be there is an improvement but it's obviously not too drastic, the main Ardour canvas is still quite visibly lagging when the analyzers are on and is redrawn smoothly when they are off. The idea with cairo_push_group and cairo_pop_group is definitely worth a try, may be different drivers behave differently when it comes to initialization/uninitialization of hardware resources.

**Comment by derkrasseleo** [comment]
> I've just updated cairo surfaces to use `cairo_surface_create_similar` instead of `cairo_image_surface_create`. Also I changed antialiasing mode from GOOD to FAST and pushed changes to `devel` branch. @Efenstor @wargreen @derkrasseleo are you available to build from `devel` branch and test if something has changed?

I tried building from `devel` branch but got this error at the end of the build:
```
cp: regular file '/usr/local/lib/pkgconfig/lsp-r3d-glx-lib.pc' cannot be created: No permission
make[5]: *** [Makefile:165: install] Error 1
make[5]: Directory "/home/leo/documents/git/lsp-plugins/modules/lsp-r3d-glx-lib/src" is left
make[4]: *** [Makefile:53: install] Error 2
make[4]: Directory "/home/leo/documents/git/lsp-plugins/modules/lsp-r3d-glx-lib" is exited
make[3]: *** [Makefile:1140: install_LSP_R3D_GLX_LIB] Error 2
make[3]: Directory "/home/leo/documents/git/lsp-plugins/modules/lsp-plugin-fw/src" is exited
make[2]: *** [Makefile:67: install] Error 2
make[2]: Directory "/home/leo/documents/git/lsp-plugins/modules/lsp-plugin-fw" is exited
make[1]: *** [Makefile:145: install] Error 2
make[1]: Directory "/home/leo/documents/git/lsp-plugins/src" is exited
make: *** [Makefile:68: install] Error 2
```

**Comment by sadko4u** [comment]
@derkrasseleo you need root privileges to run `make install`.

**Comment by derkrasseleo** [comment]
> @derkrasseleo you need root privileges to run `make install`.

I've built them successfully but haven't noticed any change in framerate so far.

**Comment by Efenstor** [comment]
In the light of preparing to do the final mix of my latest album I decided to finally figure out whether or not I can use LSP plugins, because they are way too cool to disregard because I had had problems with them. I spent almost three days experimenting with different kernels, different kernel settings (mostly regarding C-States), BIOS settings, GPU settings, tried Clang 14 and Clang 19 for LSP and for Ardour, tried the native GTK and the Ardour's internal YTK with no conclusive results, sometimes it seemed that it did some improvement to the Ardour's UI responsiveness, sometimes it did nothing at all.

Until I tried VST3 instead of LV2. Immediately I saw a huge improvement. I thought to myself: no way! But as soon as I placed Multiband Limiter Stereo to the Master track, its UI appeared but the Ardour's UI completely froze. Funnily enough if I scrolled the mouse button over the non-drawn area the UI appeared but the track order was messed up and otherwise nothing else responded to mouse clicks, yet I could play and stop the project using the Space key.

Then today I finally remembered that there is one more option that I did not try: VST2. And that was it! It worked perfectly, and now, just as I hoped, the analyzers in the plugins are redrawn dropping frames, while the Ardour's UI is 100% smooth. 

Basically the only thing remains to figure out what's wrong with the VST3 version, and why almost no one else experiencing the same problem with the LV2 version (which still has the benefit of the inline displays).

My hardware configuration:

- AMD Ryzen 7 2700 Eight-Core Processor
- AMD Radeon RX 5600 XT
- 2160p HIDPI monitor via DisplayPort
- HVME Drive

My software configuration:

- Debian 12 (Bookworm) stable
- Kernel: 6.11.10+bpo-amd64 (from backports)
- Ardour 8.10 compiled from the source code
- LSP plugins git master compiled from the source code

Ardour was compiled with `./waf configure --freedesktop --optimize --docs`
LSP plugins were compiled with `make config FEATURES="lv2 vst2 doc ui"`

**Comment by sadko4u** [comment]
I think the main reason is `libcairo` which is slow at drawing line joints.
I suppose we finally need to move towards OpenGL vector graphics implementation instead of Cairo one.

**Comment by brummer10** [comment]
As I said in my previous comment, using  
`cairo_push_group (cr); `
and , when drawing is finished, 
```
cairo_pop_group_to_source (cr); 
cairo_paint (cr); 
```
makes a huge different. 

>Nearly 100% of the time, when someone's fallen into a slow path using cairo, it's because they've managed to get cairo to copy something back from the graphics server, do some operations on the CPU, and then reupload back to the server. And surprise, that's very slow.

keep in mind that every call to cairo_stroke() or, fill or paint, call the graphic server. That could be avoided by using push and pop. Just pop the complete drawing at once to the server and do all drawings offline.

**Comment by sadko4u** [comment]
This doesn't matter if we draw long line of 640 segments. It's a single `cairo_stroke` call. But it can completely slow down graphics.

**Comment by brummer10** [comment]
Alternative you could move the rendering in a background thread with a reasonable timeout so that you didn't block the host GUI thread.

**Comment by sadko4u** [comment]
Hi all!
I've just pushed all changes related to OpenGL rendering to the `devel` branch.
I would like to ask everyone intereseted to build plugins from the `devel` branch and test the performance of plugin window with OpenGL rendering.
To compare the performance with Cairo, you can disable OpenGL rendering by setting up envrionment variable `LSP_WS_LIB_GLXSURFACE=off` and restart plugins.

**Comment by Efenstor** [comment]
> Hi all! I've just pushed all changes related to OpenGL rendering to the `devel` branch. I would like to ask everyone intereseted to build plugins from the `devel` branch and test the performance of plugin window with OpenGL rendering. To compare the performance with Cairo, you can disable OpenGL rendering by setting up envrionment variable `LSP_WS_LIB_GLXSURFACE=off` and restart plugins.

Tried it, getting a crash once I double-click any LSP plugin to show its UI (a plugin itself can be added to a track without an issue). Ran Ardour with gdb (`ardour8 -g`), here is the useful part of the output:

```
ardour-8.10.0: fccache.c:808: FcCacheFini: Assertion `fcCacheChains[i] == NULL' failed.

Thread 1 "ArdourGUI" received signal SIGABRT, Aborted.
__pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, 
    no_tid=no_tid@entry=0) at ./nptl/pthread_kill.c:44
44	./nptl/pthread_kill.c: No such file or directory.
```

**Comment by Efenstor** [comment]
Just searched and found that fccache.c is related to libfontconfig. May be I should try cleaning its cache manually somehow?

**Comment by sadko4u** [comment]
Oh, the problem with `libfontconfig` came unexpectedly.
Also I found this ticket: https://gitlab.freedesktop.org/fontconfig/fontconfig/-/issues/227.
It seems that calling `FcFini` is not a good option when working as a plugin.
Removed the call.
Please update your 'devel' build with all dependencies and try agan.

**Comment by Efenstor** [comment]
It works! But... Incredibly enough, the problem with the Ardour interface not being redrawn is still present! I could not believe my eyes. Why no difference at all? And the behavior is 100% the same: partial freeze with LV2, total freeze with the VST3, and tolerable performance with VST2.

The plugins' UI is now clearly drawn by the GPU, I checked it using the radeontop utility, with and without LSP_WS_LIB_GLXSURFACE=off. But somehow nothing has become smoother, and if with LSP_WS_LIB_GLXSURFACE=off I have 250% load on CPU, with it's LSP_WS_LIB_GLXSURFACE=on I have 100% on the GPU in addition to the CPU. Are you sure the old code is not executing now?

**Comment by Efenstor** [comment]
May be just lowering the UI thread priority could help a bit in my case?

**Comment by Efenstor** [comment]
I tested it mostly with the Multiband Limiter plugin, now I tried Compressor Stereo and its UI has clearly become smoother and not CPU nor GPU is overloaded. But once I try to open the Multiband Limiter it freezes up Ardour almost completely, while if its UI is not open DSP load is about 30%, so it's not the problem with my CPU being too low-end for it.

**Comment by sadko4u** [comment]
> It works! But... Incredibly enough, the problem with the Ardour interface not being redrawn is still present! I could not believe my eyes. Why no difference at all? And the behavior is 100% the same: partial freeze with LV2, total freeze with the VST3, and tolerable performance with VST2.

As far as I remember, I implemented event loop in a separate thread for VST2 on Linux. That was required to make OBS studio properly working with VST2. That may be a reason why Ardour works better with VST2.
Still, your Ryzen 2700 is not a bad CPU and my current laptop's Ryzen 4800H with built-in graphics doesn't outperform it too much. I suppose the main reason why Ardour is freezing is that main loop (which is also responsible for drawing) is utilized at 100%.
I hoped that the performance with OpenGL will significantly raise (and I observe good results on my machine) but I don't have HiDPI display.

> But somehow nothing has become smoother, and if with LSP_WS_LIB_GLXSURFACE=off I have 250% load on the CPU, with LSP_WS_LIB_GLXSURFACE=on I have 100% on the GPU in addition to the CPU.

You can check the usage of OpenGL or Cairo by turning on the logs. This can be done at configuration stage:
```
make config ... TRACE=1
```
Plugins will write logs to the `/tmp` directory:

* `/tmp/lsp-jack-lib.log` for JACK
* `/tmp/lsp-vst2-aeffect.log` for VST2
* `/tmp/lsp-vst3.log` for VST3
* `/tmp/lsp-lv2ui.log` for LV2 UI

You can see the following message there if UI uses OpenGL:
```
create_surface: Using X11GLSurface ptr=<address>
```
and the following message if UI uses Cairo:
```
create_surface: Using X11CairoSurface ptr=<address>
```
To ensure what actual graphics driver is used for plugin window.

**Comment by sadko4u** [comment]
OK, I was able to run all my OpenGL stuff in the [renderdoc tool](https://renderdoc.org/) and found some serious performance issues related to drawing graphics.
I've optimized graphics to minimize number of draw calls to OpenGL. Still not perfect but significantly lowered.
Please update all stuff to actual `devel` branch and try again.

**Comment by wargreen** [comment]
> It works! But... Incredibly enough, the problem with the Ardour interface not being redrawn is still present! I could not believe my eyes. Why no difference at all? And the behavior is 100% the same: partial freeze with LV2, total freeze with the VST3, and tolerable performance with VST2.

Maybe i'm missing the point, but as far i remember i have this issue with all opengl plugins when i have more than 1 GUI open. All the Ardour's GUI is frozen and come back when i close the GUI.
Maybe an issue with wayland, on my two computers.

**Comment by sadko4u** [comment]
I also want to introduce some text rendering optimizations, so please keep being in touch to test it.

**Comment by Efenstor** [comment]
> I also want to introduce some text rendering optimizations, so please keep being in touch to test it.

Of course! Thank you so much for your efforts!

**Comment by sadko4u** [comment]
OK, I've added rendering of fonts using single shared texture.
It additionally allowed me to reduce the overall number of draw calls when rendering text data.
Please update and try.

**Comment by Efenstor** [comment]
> OK, I've added rendering of fonts using single shared texture. It additionally allowed me to reduce the overall number of draw calls when rendering text data. Please update and try.

Tested it: works perfectly, I didn't notice any anomalies in any of the texts, and the anti-aliasing is great. I didn't measure the performance of the previous version so I can't compare if it has become any smoother. Anyway it works great to me :)

**Comment by sadko4u** [comment]
Please don't close tickets until they get a release.

**Comment by Efenstor** [comment]
Hold on: after about of hour of practical mixing work with different plugins (mostly Compressor and Parametric Equalizer) I've noticed a weird behavior with the GPU load: even when all the UI's are closed the video card's fans are still running full-speed and the radeontop utility shows the 100% load. If I close Ardour and re-open the project the GPU load gets back to normal. I am not a programmer, but is everything that should be deallocated is really deallocated when the UI is closed? All the surfaces or whatever is used?

**Comment by Efenstor** [comment]
Yes indeed, if I just open and close the UI's of several different plugin instances I quickly run out of VRAM. Opening and closing the same instance does not cause the issue.

Here is how it looks right after opening the session:

![Image](https://github.com/user-attachments/assets/23f005e8-1583-454e-80d5-2e27f0281ca3)

And here is after opening and closing the plugin interfaces:

![Image](https://github.com/user-attachments/assets/bc1693fc-76cc-4876-a193-7c8f5e54f8c7)

**Comment by sadko4u** [comment]
I know that Ardour is not destroying plugin UI windows when hiding them. This may be a reason. I mean, when window is deallocated, all widget surfaces still continue exist. And this can be a reason why you get out of the VRAM.

**Comment by Efenstor** [comment]
I've found now that if I close the session and open another one VRAM does not get freed, not even a bit, but only if I close Ardour itself. That's strange. Anyway I guess at least something should be done about VRAM overflow because a complex session may contain dozens of plugins.

**Comment by Efenstor** [comment]
Wow, I didn't know that a lot of spears have already been broken about this very topic: https://discourse.ardour.org/t/very-high-gpu-usage-on-windows/107672

**Comment by Efenstor** [comment]
I guess this is the solution:

![Image](https://github.com/user-attachments/assets/29b39c82-a156-4ca7-802d-283ffd5525b9)

**Comment by Efenstor** [comment]
I tested it a bit and VRAM is still being kept filled to a significant degree after the UI's were closed, yet the graphics pipe load now drops to zero which is great. I am not an expert, may be that thing with VRAM is normal, here is the discussion for example: https://stackoverflow.com/questions/39784072/is-there-garbage-collection-on-the-gpu

**Comment by sadko4u** [comment]
@Efenstor 
OK, I've changed OpenGL resource allocation mechanism. You can update now.

Still we met another problems on devices based on NVIDIA cards and I need to correct them.

**Comment by Efenstor** [comment]
Tested the latest update. I still observe the gradual accumulation of the VRAM usage, even with the "Destroy GUI instance" option on. Actually this option is quite helpful in debugging this issue because you can open and close the same plugin a number of times. Once the UI is closed the VRAM usage falls back but only to about 50% of the previous level, and it happens until all the VRAM is exhausted; once it's full I begin to observe various lags, for example new LSP plugin windows appear only after several seconds after I the double click the button, and if I open Firefox even the mouse cursor begins to freeze for some moments.

**Comment by Efenstor** [comment]
Just a moment, may be I am wrong, I did not do `make clean` the previous time, only `make fetch` and `make`, now I did it and now the behavior seems improved: the VRAM usage reached 85-95% and always gets back to these numbers, no matter how many times I open and close any plugin window.

**Comment by Efenstor** [comment]
Nope, that's still a problem: when VRAM is 90-95% I get a lot of audio dropouts and DSP indicator gets into the red zone all the time. If I just close Ardour and open the same session anew the DSP load is shown only 25% instead of 90-95%.

**Comment by sadko4u** [comment]
I think it is a good reason to reboot your computer before doing new tests as some video drivers do not free allocated memory. I met this when doing my first steps with OpenGL.
Anyway, I'm testing standalone versions of plugins now. It seems I need to check the behaviour directly in Ardour.

**Comment by Efenstor** [comment]
Tried rebooting, sadly no improvement.

Here is how it looks after just opening and closing the Multiband Limiter UI about 20 times. And the DSP load begins to erratically jump from reasonable values to 100%.

![Image](https://github.com/user-attachments/assets/0e9abdb1-73a3-4d01-91b0-d50de11372d7)

**Comment by sadko4u** [comment]
@Efenstor you're testing on VST2 plugin format, right?

First, please update. I've again changed some code, so you need to fetch latest changes to the `devel` branch.

Second, I see that Ardour doesn't unmap plugin's window when hiding it. It just hides parent window, and that's all. So we don't catch the X11 window unmap notification and don't release OpenGL context of the plugin's window and all related data. That's very sad because we have no mechanism at this moment to free allocated data.

@x42 do we have some way to detect that Ardour has hidden plugin's window without destroying it?

The only one thing to free allocated data is adding garbage collector which will free cached off-screen surfaces if they weren't updated for a long time (second or more).
As I see, the LV2 implementation doesn't allow to hide plugin windows and my version of Ardour always destroys the plugin's window. Did you try the LV2 version of plugins?

**Comment by x42** [comment]
> [@x42](https://github.com/x42) do we have some way to detect that Ardour has hidden plugin's window without destroying it?

In general it is good practice to just check if the window is mapped, regardless of host or OS, and stop updates when it's offscreen.

**Comment by sadko4u** [comment]
@x42
>In general it is good practice to just check if the window is mapped, regardless of host or OS, and stop updates when it's offscreen.

The problem here is that I don't see in logs that I receive unmap message when plugin's window is closed. Seems that parent window is hiding without unmapping it's child.

**Comment by sadko4u** [comment]
@Efenstor do I right understand that you still have problems with VRAM even on LV2?

**Comment by Efenstor** [comment]
Yes, correct. Tested it once again to be sure, on the 11th opening of the LSP Multiband Limiter (LV2) window I started to have visible lags, checked with radeontop and indeed the NVRAM was 90% full.

**Comment by sadko4u** [comment]
@Efenstor OK, seems that I found the reason. I tried to delete active OpenGL context and this function failed. That's why OpenGL context was left unfreed until the application was closed. Now closing the UI window drops the GTT use back to normal value. Please update and try.

I hope the only problem left is that Ardour does not unmap windows when closing plugin's window in a default plugin window operating mode.

**Comment by Efenstor** [comment]
Success!!! 💯

VRAM usage now drops back to the original value every time. Strange but even with the "Only hides the window" setting in the Preferences there are no memory or load problems whatsoever, so may be there is nothing to fix with the unmapping? I opened and closed various plugins (their UI windows) exactly 50 times, same plugins and different plugins on different tracks: GPU load, VRAM usage and CPU always get back to the original values. I tried both LV2 and VST2, the behavior is identically good (VST2 consumes even less resources, I guess that's due to the lack of inline displays).

**Comment by sadko4u** [comment]
Fixed in 1.2.21!
